c/client.c.go

```diff
@@ -0,0 +1,223 @@
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	code, status, err := client.SendText(ctx, C.GoString(msgC))
+	if err != nil {
+		log.Printf("%v\n", err)
+		return int(codes.ERR_SEND_TEXT)
+	}
+	fmt.Printf("code returned: %s\n", code)
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientSendFile
+func ClientSendFile(ctxC unsafe.Pointer, clientPtr uintptr, fileName *C.char, length C.int, fileBytes unsafe.Pointer, codeOutC **C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	// TODO: is there a way to avoid copying?
+	reader := bytes.NewReader(C.GoBytes(fileBytes, length))
+
+	code, status, err := client.SendFile(ctx, C.GoString(fileName), reader, false)
+	if err != nil {
+		return int(codes.ERR_SEND_TEXT)
+	}
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientRecvText
+func ClientRecvText(ctxC unsafe.Pointer, clientPtr uintptr, codeC *C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	go func() {
+		msg, err := client.Receive(ctx, C.GoString(codeC), false)
+
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT))
+		}
+
+		data, err := ioutil.ReadAll(msg)
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT_DATA))
+		}
+		dataC := C.CBytes(data)
+		fileC := (*C.file_t)(C.malloc(C.sizeof_file_t))
+		*fileC = C.file_t{
+			length: C.int(len(data)),
+			data:   (*C.uint8_t)(dataC),
+		}
+
+		C.call_callback(ctxC, cb, unsafe.Pointer(fileC), C.int(codes.OK))
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientRecvFile
+func ClientRecvFile(ctxC unsafe.Pointer, clientPtr uintptr, codeC *C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	go func() {
+		msg, err := client.Receive(ctx, C.GoString(codeC), false)
+
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT))
```

Comment:
I'm realizing the error codes for send/recv file are wrong. They don't exist yet.